File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        m, n = len(str1), len(str2)
        cur = ""
        res = cur
        for i in range(min(m, n)):
            cur += str1[i]
            # print(cur)

            if m % (i + 1) != 0 or n % (i + 1) != 0:
                continue

            if cur * (m // (i + 1)) == str1 and cur * (n // (i + 1)) == str2:
                res = cur

        return res

File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        for (int i = Math.min(str1.length(), str2.length()); i > 0; i--) {
            if (valid(str1, str2, i)) return str1.substring(0, i);
        }
        return "";
    }

    public boolean valid(String str1, String str2, int i) {
        int l1 = str1.length(), l2 = str2.length();
        if ((l1 % i != 0) && (l2 % i != 0)) return false;
        String base = str1.substring(0, i);
        return str1.replace(base, "").isEmpty() && str2.replace(base, "").isEmpty();
    }
}